neuralmonkey/encoders/raw_rnn_encoder.py

```diff
@@ -0,0 +1,178 @@
+                CoverageAttention (default 3).
+        """
+        ModelPart.__init__(self, name, save_checkpoint, load_checkpoint)
+        Attentive.__init__(
+            self, attention_type, attention_fertility=attention_fertility)
+
+        assert check_argument_types()
+
+        self.data_id = data_id
+
+        self.rnn_size = rnn_size
+        self.max_input_len = max_input_len
+        self.input_dimension = input_dimension
+        self.dropout_keep_p = dropout_keep_prob
+        self.use_noisy_activations = use_noisy_activations
+        self.parent_encoder = parent_encoder
+
+        log("Initializing RNN encoder, name: '{}'"
+            .format(self.name))
+
+        with tf.variable_scope(self.name):
+            self._create_input_placeholders()
+
+            self._input_mask = tf.sequence_mask(self._input_lengths,
+                                                dtype=tf.float32)
+
+            fw_cell, bw_cell = self.rnn_cells()  # type: RNNCellTuple
+            outputs_bidi_tup, encoded_tup = tf.nn.bidirectional_dynamic_rnn(
+                fw_cell, bw_cell, self.inputs, self._input_lengths,
+                dtype=tf.float32)
+
+            self.hidden_states = tf.concat(outputs_bidi_tup, 2)
+
+            with tf.variable_scope('attention_tensor'):
+                self.__attention_tensor = self._dropout(
+                    self.hidden_states)
+
+            self.encoded = tf.concat(encoded_tup, 1)
+
+        log("RNN encoder initialized")
+
+    @property
+    def _attention_tensor(self):
+        return self.__attention_tensor
+
+    @property
+    def _attention_mask(self):
+        return self._input_mask
+
+    def _create_input_placeholders(self):
+        """Creates input placeholder nodes in the computation graph"""
+        self.train_mode = tf.placeholder(tf.bool, shape=[],
+                                         name="mode_placeholder")
+
+        self.inputs = tf.placeholder(tf.float32,
+                                     shape=[None, None,
+                                            self.input_dimension],
+                                     name="encoder_input")
+
+        self._input_lengths = tf.placeholder(
+            tf.int32, shape=[None],
+            name="encoder_padding_lengths")
+
+    def _dropout(self, variable: tf.Tensor) -> tf.Tensor:
+        """Perform dropout on a variable
+
+        Arguments:
+            variable: The variable to be dropped out
+
+        Returns:
+            The dropped value of the variable
+        """
+        if self.dropout_keep_p == 1.0:
+            return variable
+
+        # TODO as soon as TF.12 is out, remove this line and use train_mode
+        # directly
+        train_mode_batch = tf.fill(tf.shape(variable)[:1], self.train_mode)
+        dropped_value = tf.nn.dropout(variable, self.dropout_keep_p)
+        return tf.where(train_mode_batch, dropped_value, variable)
+
+    def rnn_cells(self) -> RNNCellTuple:
```

Comment:
Nestálo by za to, aby tohle byla base class pro sentence encoder? Jestli se ti to nechce dělat, já to klidně udělám potom, co se tohle zmerguje.